Make flare gun explode, damage and stun the player when used with the wrong ammo

MendaxxDev opened the pull request:
## Why / Balance
To make the flare gun more realistic, and also to actually make it different from a shotgun.

**Changelog**
:cl:
- tweak: The flare gun now explodes, damages and stuns the player when used with wrong ammo

Kadeo64 commented:
Finally, the flare gun is no longer a concealable pocket cannon.

deltanedas commented:
like a comment i saw, perhaps let it shoot all shells but anything that isnt a flare explodes the gun (and does some dmg)

IProduceWidgets commented:
Does this also prevent them from being loaded into regular shotguns?

Emisse commented:
i agree with delta but if thats out of scope for your ability this is fine

Doru991 commented:
Good PR, the improvised shotgun is currently 100% useless since random flare guns you find in maints already do everything it does but better so this gives it a reason to exist

MendaxxDev commented on Resources/Prototypes/Entities/Objects/Weapons/Guns/Ammunition/Cartridges/shotgun.yml:
```diff
@@ -57,6 +57,9 @@
   components:
+  - type: Tag
+    tags:
+      - ShellShotgunFlare
```
Mhh the tag is `ShellShotgunFlare`, I don't think it should make things different from how they are right now. I can't test it at the moment though.

MendaxxDev commented:
> like a comment i saw, perhaps let it shoot all shells but anything that isnt a flare explodes the gun (and does some dmg)

Okay, I've changed things and implemented this. Don't know if how I've done it is the best way, but this approach should be general enough to make it possible to do the same thing with other guns/ammo if one wants.

deltanedas commented on Content.Server/Weapons/Ranged/Systems/GunSystem.cs:
```diff
@@ -132,6 +132,21 @@ private void OnBallisticPrice(EntityUid uid, BallisticAmmoProviderComponent comp
+                            !gun.CompatibleAmmo.Exists(ammoAllowed => ammoAllowed.Equals(cartridge.Prototype))
+                            && user != null)
+                        {
+                            Damageable.TryChangeDamage(user, new DamageSpecifier(), origin: user);
```
what

deltanedas commented on Content.Server/Weapons/Ranged/Systems/GunSystem.cs:
```diff
@@ -132,6 +132,21 @@ private void OnBallisticPrice(EntityUid uid, BallisticAmmoProviderComponent comp
+
+                            Audio.PlayPvs(new SoundPathSpecifier("/Audio/Weapons/Guns/Gunshots/bang.ogg"), gunUid);
+
+                            PopupSystem.PopupEntity(Loc.GetString("Test"), user.Value);
```
:trollface:

deltanedas commented on Content.Shared/Weapons/Ranged/Components/GunComponent.cs:
```diff
@@ -110,6 +112,13 @@ public partial class GunComponent : Component
+    /// <summary>
+    /// Type of ammo the gun can work with
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite),
+     DataField("compatibleAmmo", customTypeSerializer: typeof(PrototypeIdListSerializer<TagPrototype>))]
+    public List<string>? CompatibleAmmo = null;
```
```cs
    [ViewVariables(VVAccess.ReadWrite), DataField]
    public List<ProtoId<TagPrototype>>? CompatibleAmmo;
```

deltanedas commented on Content.Server/Weapons/Ranged/Systems/GunSystem.cs:
```diff
@@ -132,6 +133,28 @@ private void OnBallisticPrice(EntityUid uid, BallisticAmmoProviderComponent comp
+                            !gun.CompatibleAmmo.Exists(ammoAllowed => ammoAllowed.Equals(cartridge.Prototype))
+                            && user != null)
+                        {
+                            var damage = new DamageSpecifier();
```
either put this in the component yaml for flare gun, or do something *like* this seen in other parts of the codebase (forgor the exact syntax)
```cs
var damage = new DamageSpecifier()
{
    ("Blunt", 6)
};
```

MendaxxDev replied:
Done. In this way one could also implement it for other guns, but can also totally ignore it since it's nullable.

deltanedas commented on Content.Server/Weapons/Ranged/Systems/GunSystem.cs:
```diff
@@ -132,6 +133,28 @@ private void OnBallisticPrice(EntityUid uid, BallisticAmmoProviderComponent comp
+                            !gun.CompatibleAmmo.Exists(ammoAllowed => ammoAllowed.Equals(cartridge.Prototype))
+                            && user != null)
+                        {
+                            var damage = new DamageSpecifier();
```
this line no longer needed

deltanedas commented on Content.Server/Weapons/Ranged/Systems/GunSystem.cs:
```diff
@@ -132,6 +133,28 @@ private void OnBallisticPrice(EntityUid uid, BallisticAmmoProviderComponent comp
+
+                            Audio.PlayPvs(new SoundPathSpecifier("/Audio/Weapons/Guns/Gunshots/bang.ogg"), gunUid);
+
+                            PopupSystem.PopupEntity(Loc.GetString("Wrong ammo!"), user.Value);
```
loc.getstring should be passed a locale id which is stored in Resources/Locale/...

if the ftl looks like `gun-component-wrong-ammo = Wrong ammo!` then use `Loc.GetString("gun-component-wrong-ammo")`

deltanedas commented on Content.Shared/Weapons/Ranged/Components/GunComponent.cs:
```diff
@@ -110,6 +114,18 @@ public partial class GunComponent : Component
+    /// <summary>
+    /// Type of ammo the gun can work with
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite), DataField("compatibleAmmo")]
```
just do `[ViewVariables(VVAccess.ReadWrite), DataField]` it will set tag automatically

MendaxxDev replied:
Not sure what's the problem, but if I do it like this I can't load ammo into the flare gun anymore

deltanedas replied:
sus, otherwise this lgtm

MendaxxDev replied:
Made some more tests but yeah, it stops working. I notice that every other data field in `GunComponent.cs` is declared in this way, I don't know if there's a reason for that.

deltanedas replied:
its just old way
if the tag is the same as the field name with first letter lowercase it should work

not that important anyway

deltanedas commented on Content.Shared/Weapons/Ranged/Components/GunComponent.cs:
```diff
@@ -110,6 +114,18 @@ public partial class GunComponent : Component
+    /// <summary>
+    /// Damage the gun deals when used with wrong ammo
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite), DataField("damageOnWrongAmmo")]
```
above

superjj18 commented:
Id say the stun is unnecessary, just destroy the gun and give a little bit of burn damage
